### PR TITLE
🚨 Bloom support for cache class

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Follow-up from #31421, adds support for Cache class in Bloom models. This is a breaking change as the cache class currently has its keys in shape=`(bs*heads, seq_len, dim)` as opposed to prev shape=`(bs*heads, dim, seq_len)`

This change was needed because cache class assumes that concatenation is always performed along "dim=-2"

### CODE REVIEW

**Review by gante — COMMENTED** [review]
In general it looks like a move in the right direction! Thank you for moving these changes into a separate PR 🤗 

Since we're breaking BC on the BLOOM cache, I'd like to suggest to go all the way: to make the cache here have the standard shape of `(batch_size, num_key_value_heads, seq_len, head_dim)`, even if it requires touching some modeling code and/or reshapes. If we can do so, then:
1. BLOOM would be compatible with all caches (in the current version, it is not compatible with e.g. StaticCache)
2. we could remove `_convert_to_standard_cache`, `_convert_to_bloom_cache`, and all related code in `generate`, after a deprecation cycle 💛 

Finally, regardless of the comment above, we should provide full BC during the deprecation cycle.

WDYT?

**Comment by zucchini-nlp** [comment]
Exactly, I was also thinking of that but didn't want to break Bloom more. Will add full compatibility with Cache then

**Comment by gante** [comment]
(pls ping when it's ready for a re-review!)

**Comment by zucchini-nlp** [comment]
@gante ready for review

- added cache position
- deprecation warning for legacy cache until then next release
- bloom has standard format cache and some bloom tests aren't skipped anymore

**Comment by zucchini-nlp** [comment]
@gante ping

File: src/transformers/models/bloom/modeling_bloom.py
Comment (gante) [inline src/transformers/models/bloom/modeling_bloom.py:837]: I think we can add a `#copied from ... llama` on this one. The base structure of this function was updated, but the copy statement will take care of it!

File: src/transformers/models/bloom/modeling_bloom.py
Comment (gante) [inline src/transformers/models/bloom/modeling_bloom.py:949]: not sure if possible: add a copy statement to this class (from llama)?

**Review by gante — APPROVED** [review]
Yes! One fewer special case 🔥 

In addition to the comments: have you confirmed that related slow tests are passing?

File: src/transformers/generation/utils.py
Comment (amyeroberts) [inline src/transformers/generation/utils.py:650]: The standardize_cache_format arg is now no longer needed by the function

File: src/transformers/generation/utils.py
Comment (amyeroberts) [inline src/transformers/generation/utils.py:3016]: Is this right? I'm pretty sure the cache types still need to be a tuple

File: src/transformers/models/bloom/modeling_bloom.py
Comment (amyeroberts) [inline src/transformers/models/bloom/modeling_bloom.py:207]: Let's use a better name here "shape" is a standard key word used by jax, tf, pytorch etc. to get the shape of a tensor. `_reshape` would be more apt

File: src/transformers/models/bloom/modeling_bloom.py
Comment (amyeroberts) [inline src/transformers/models/bloom/modeling_bloom.py:174]: nit

```suggestion
    def __init__(self, config: BloomConfig, layer_idx: Optional[int] = None):
```

File: src/transformers/models/bloom/modeling_bloom.py
Comment (amyeroberts) [inline src/transformers/models/bloom/modeling_bloom.py:268]: nit 
```suggestion
        # 3 x [batch_size, num_heads, seq_length, head_dim]
        query_layer, key_layer, value_layer = self._shape(fused_qkv)
```

File: src/transformers/models/bloom/modeling_bloom.py
Comment (amyeroberts) [inline src/transformers/models/bloom/modeling_bloom.py:367]:
```suggestion
    def __init__(self, config: BloomConfig, layer_idx: Optional[int] = None):
```

File: src/transformers/models/bloom/modeling_bloom.py
Comment (amyeroberts) [inline src/transformers/models/bloom/modeling_bloom.py:642]: We're already at 4.43

File: src/transformers/models/bloom/modeling_bloom.py
Comment (amyeroberts) [inline src/transformers/models/bloom/modeling_bloom.py:632]:
```suggestion
        if self.gradient_checkpointing and self.training and use_cache:
            logger.warning_once(
                "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
            )
            use_cache = False
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

+1 to @gante's comments about copied from

File: src/transformers/models/bloom/modeling_bloom.py
Comment (zucchini-nlp) [reply src/transformers/models/bloom/modeling_bloom.py:641]: oops, the PR was made hoping we'll make it to release. I'll leave one major release in between and remove it in 4.45 then

File: src/transformers/generation/utils.py
Comment (amyeroberts) [reply src/transformers/generation/utils.py:3016]: I'm talking about the change in the isinstance call, I'm pretty sure it just takes 2 arguments

```py
In [3]: isinstance("a", str, int)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 isinstance("a", str, int)

TypeError: isinstance expected 2 arguments, got 3

In [4]: isinstance("a", (str, int))
Out[4]: True
```

Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:3016]: oh right, missed that one!

File: src/transformers/models/bloom/modeling_bloom.py
Comment (zucchini-nlp) [reply src/transformers/models/bloom/modeling_bloom.py:1086]: Bloom has `head_mask` and doesn't use `positiion_ids`, so can't be copied. Same for other copy statments

**Comment by zucchini-nlp** [comment]
Couldn't add copy statements because Bloom is different from Llama, and accepts some other kwargs. Addressed all other comments

**Comment by LysandreJik** [comment]
Maybe a final review from @ArthurZucker would be nice here

**Comment by zucchini-nlp** [comment]
Yes, this one and the other PR that deprecates cache will be waiting for @ArthurZucker :)

File: src/transformers/models/bloom/modeling_bloom.py
Comment (ArthurZucker) [inline src/transformers/models/bloom/modeling_bloom.py:275]: we are transposing twice instead of a permute, but alright!

File: src/transformers/models/bloom/modeling_bloom.py
Comment (ArthurZucker) [inline src/transformers/models/bloom/modeling_bloom.py:278]: this should be cache positions no?

File: src/transformers/models/bloom/modeling_bloom.py
Comment (ArthurZucker) [reply src/transformers/models/bloom/modeling_bloom.py:836]: should we add copied from here?

**Review by ArthurZucker — APPROVED** [review]
Very nice cleanup

File: src/transformers/models/bloom/modeling_bloom.py
Comment (zucchini-nlp) [reply src/transformers/models/bloom/modeling_bloom.py:278]: Yes, we can do cache position instead. 

And for some reason I can't reply to the below comment, so I'll do it here: no, we can'y copy because then we'll be passing position ids to forward and trigger deprecation warning every `generate()`

Comment (zucchini-nlp) [reply src/transformers/models/bloom/modeling_bloom.py]: done!

**Comment by zucchini-nlp** [comment]
Will merge it today evening, got approve from one of the core maintainers. @amyeroberts , let me know if you want to review once more, so I won't merge yet :)

**Pull request merged**
